### ISSUE TITLE
CI: K8sServiceTest Tests NodePort BPF

### ISSUE THREAD

**Issue by brb** [cilium/cilium#8945]
The test suite seems to be flaky. The example failure:

```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/k8s-1.10-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:420
Pod "testclient-6zb5g" can not connect to service "http://127.0.0.1:32174"
Expected command: kubectl exec -n default testclient-6zb5g -- curl --path-as-is -s -D /dev/stderr --fail --connect-timeout 5 --max-time 8 http://127.0.0.1:32174 -w "time-> DNS: '%{time_namelookup}(%{remote_ip})', Connect: '%{time_connect}',Transfer '%{time_starttransfer}', total '%{time_total}'" 
To succeed, but it failed:
Exitcode: 28 
Stdout:
 	 time-> DNS: '0.000021()', Connect: '0.000000',Transfer '0.000000', total '5.000594'
Stderr:
 	 command terminated with exit code 28 
```

The following instances of the failure were observed:

- From pod curl to ClusterIP.
- From pod curl to 127.0.0.1:NodePort
- From pod curl to PublicIP:NodePort

The tunnel / direct routing mode setting does not matter, as the test failed in both modes. In all cases `bpf_sock.c` (aka host-reachable services) was involved.

* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/14436/testReport/junit/Suite-k8s-1/10/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan/ : 
[de1ae283_K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan.zip](https://github.com/cilium/cilium/files/3516090/de1ae283_K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan.zip)

* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/14440/testReport/junit/Suite-k8s-1/10/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan/ : 
[d2d95800_K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan.zip](https://github.com/cilium/cilium/files/3515203/d2d95800_K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_vxlan.zip)

/cc @borkmann

**Comment by brb** [comment]
Drop counts from nodes which were running the client pods:

```
direction="EGRESS" reason="Invalid source ip" 33.000000
direction="EGRESS" reason="Unknown L3 target address" 2.000000
direction="EGRESS" reason="Unsupported protocol for NAT masquerade" 3.000000
```
```
reason="Invalid source ip" direction="EGRESS" 65.000000
direction="EGRESS" reason="Missed tail call" 2.000000
direction="EGRESS" reason="Unknown L3 target address" 17.000000
direction="EGRESS" reason="Unsupported protocol for NAT masquerade" 1.000000
direction="INGRESS" reason="Policy denied (L3)" 12.000000
```

```
direction="EGRESS" reason="Invalid source ip" 134.000000
direction="EGRESS" reason="Policy denied (L3)"  1.000000
direction="EGRESS" reason="Unknown L3 target address" 15.000000
direction="EGRESS" reason="Unsupported L3 protocol" 12.000000
direction="EGRESS" reason="Unsupported protocol for NAT masquerade" 8.000000
direction="INGRESS" reason="Policy denied (L3)"  30.000000
direction="INGRESS" reason="Stale or unroutable IP"   1.000000
```

Checked locally that `curl -connect-timeout 5 --max-time 8` during the specified period of time tries to send 3 TCP SYN packets. So, if the failures were due to a drop, then a drop count should be >= 3.

**Comment by borkmann** [comment]
Based on the above, it's trying to connect to `http://127.0.0.1:32174`. From the direct routing zip file's bugtool output, the only service entries I see exposed as wild-carded are:

Cilium instance 1:
```
key: 00 00 00 00 77 e0 02 00  00 00 00 00  value: 09 00 00 00 00 00 00 03  00 00 00 00
key: 00 00 00 00 77 e0 01 00  00 00 00 00  value: 08 00 00 00 00 00 00 03  00 00 00 00
key: 00 00 00 00 77 e0 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 03  00 00 00 00
```

Cilium instance 2:
```
key: 00 00 00 00 77 e0 01 00  00 00 00 00  value: 08 00 00 00 00 00 00 04  00 00 00 00
key: 00 00 00 00 77 e0 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 04  00 00 00 00
key: 00 00 00 00 77 e0 02 00  00 00 00 00  value: 09 00 00 00 00 00 00 04  00 00 00 00
```

Hm, nothing with port `7d ae`. The `77 e0` is 30688, and given nothing might be listening in the pod at `127.0.0.1:32174`, it might just time out.

**Comment by borkmann** [comment]
> Based on the above, it's trying to connect to `http://127.0.0.1:32174`. From the direct routing zip file's bugtool output, the only service entries I see exposed as wild-carded are:
> 
> Cilium instance 1:
> 
> ```
> key: 00 00 00 00 77 e0 02 00  00 00 00 00  value: 09 00 00 00 00 00 00 03  00 00 00 00
> key: 00 00 00 00 77 e0 01 00  00 00 00 00  value: 08 00 00 00 00 00 00 03  00 00 00 00
> key: 00 00 00 00 77 e0 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 03  00 00 00 00
> ```
> 
> Cilium instance 2:
> 
> ```
> key: 00 00 00 00 77 e0 01 00  00 00 00 00  value: 08 00 00 00 00 00 00 04  00 00 00 00
> key: 00 00 00 00 77 e0 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 04  00 00 00 00
> key: 00 00 00 00 77 e0 02 00  00 00 00 00  value: 09 00 00 00 00 00 00 04  00 00 00 00
> ```
> 
> Hm, nothing with port `7d ae`. The `77 e0` is 30688, and given nothing might be listening in the pod at `127.0.0.1:32174`, it might just time out.

Scratch that thought, the correct sysdump has 30688 exposed, so that seems good.

**Comment by brb** [comment]
Managed to reproduce with `cilium monitor -D -v` being enabled. Anyway, it seems that TCP SYN sent from the client Pod (`10.10.1.86`) sent to `http://127.0.0.1:30453` via `bpf_sock` got unexpectedly SNAT'd by netfilter/iptables (`10.10.0.249` is the IP addr of the NodePort endpoint running on the other node):

```
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: Master service lookup failed
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: Conntrack lookup 1/2: src=10.10.1.86:37656 dst=10.10.0.249:80
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: Successfully mapped daddr=10.10.0.249 to identity=17742
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=41719 lb=0.0.0.0
------------------------------------------------------------------------------
CPU 01: MARK 0x27e371a2 FROM 364 to-stack: 74 bytes (74 captured), state new, identity 41719->17742
Ethernet	{Contents=[..14..] Payload=[..62..] SrcMAC=7a:ed:2c:f3:ba:e3 DstMAC=b6:ac:75:7d:bd:f6 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=21637 Flags=DF FragOffset=0 TTL=63 Protocol=TCP Checksum=53460 SrcIP=10.10.1.86 DstIP=10.10.0.249 Options=[] Padding=[]}
TCP	{Contents=[..40..] Payload=[] SrcPort=37656 DstPort=80(http) Seq=2221004874 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=64860 Checksum=5777 Urgent=0 Options=[..5..] Padding=[]}

------------------------------------------------------------------------------
CPU 01: MARK 0x27e371a2 FROM 364 DEBUG: 74 bytes, Delivery to ifindex 0
Ethernet	{Contents=[..14..] Payload=[..62..] SrcMAC=7a:ed:2c:f3:ba:e3 DstMAC=b6:ac:75:7d:bd:f6 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=21637 Flags=DF FragOffset=0 TTL=63 Protocol=TCP Checksum=53460 SrcIP=10.10.1.86 DstIP=10.10.0.249 Options=[] Padding=[]}
TCP	{Contents=[..40..] Payload=[] SrcPort=37656 DstPort=80(http) Seq=2221004874 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=64860 Checksum=5777 Urgent=0 Options=[..5..] Padding=[]}

# The following is most likely logged by bpf_netdev.o on the native device:
CPU 01: MARK 0x27e371a2 FROM 0 DEBUG: Conntrack lookup 1/2: src=192.168.36.12:37656 dst=10.10.0.249:80
```

Previously, the same Pod sent a request from the same source port to the NodePort service via the other node IP addr + NodePort port which get MASQ'd (expected due to the ` -A CILIUM_POST_nat -s 10.217.0.0/24 ! -d 10.0.0.0/8 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE` rule):

```
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: Master service lookup failed
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: Conntrack lookup 1/2: src=10.10.1.86:37656 dst=192.168.36.11:30453
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: Successfully mapped daddr=192.168.36.11 to identity=2
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=41719 lb=0.0.0.0
------------------------------------------------------------------------------
CPU 01: MARK 0x652e7696 FROM 364 to-stack: 74 bytes (74 captured), state new, identity 41719->2
Ethernet	{Contents=[..14..] Payload=[..62..] SrcMAC=7a:ed:2c:f3:ba:e3 DstMAC=b6:ac:75:7d:bd:f6 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=42248 Flags=DF FragOffset=0 TTL=63 Protocol=TCP Checksum=42656 SrcIP=10.10.1.86 DstIP=192.168.36.11 Options=[] Padding=[]}
TCP	{Contents=[..40..] Payload=[] SrcPort=37656 DstPort=30453 Seq=4227868345 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=64860 Checksum=61505 Urgent=0 Options=[..5..] Padding=[]}
------------------------------------------------------------------------------
CPU 01: MARK 0x652e7696 FROM 364 DEBUG: 74 bytes, Delivery to ifindex 0
Ethernet	{Contents=[..14..] Payload=[..62..] SrcMAC=7a:ed:2c:f3:ba:e3 DstMAC=b6:ac:75:7d:bd:f6 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..40..] Version=4 IHL=5 TOS=0 Length=60 Id=42248 Flags=DF FragOffset=0 TTL=63 Protocol=TCP Checksum=42656 SrcIP=10.10.1.86 DstIP=192.168.36.11 Options=[] Padding=[]}
TCP	{Contents=[..40..] Payload=[] SrcPort=37656 DstPort=30453 Seq=4227868345 Ack=0 DataOffset=10 FIN=false SYN=true RST=false PSH=false ACK=false URG=false ECE=false CWR=false NS=false Window=64860 Checksum=61505 Urgent=0 Options=[..5..] Padding=[]}

# Received by bpf_netdev, spot the MASQ'd sip:
CPU 01: MARK 0x652e7696 FROM 0 DEBUG: Conntrack lookup 1/2: src=192.168.36.12:37656 dst=192.168.36.11:30453
CPU 01: MARK 0x652e7696 FROM 0 DEBUG: Conntrack lookup 2/2: nexthdr=6 flags=1
CPU 01: MARK 0x652e7696 FROM 0 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x652e7696 FROM 0 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=0 lb=0.0.0.0
```

One workaround is to store cluster node IPs in ipcache. This would allow bpf_sock to do the transparent LB translation when a request sent to a NodePort SVC via other node IP addr.

**Comment by joestringer** [comment]
@brb is this still an issue? Are you tracking it?

**Comment by brb** [comment]
Yes, it's still the issue. I'm trying to triage it again.

**Comment by brb** [comment]
Cilium monitor output (direct routing, no XDP) on a node which runs a backend when we hit the packet loop case:
```
CPU 01: MARK 0x0 FROM 632 DEBUG: Attempting local delivery for container id 2553 from seclabel 6
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 1/2: src=192.168.36.12:42671 dst=10.10.0.237:69
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT entry found lifetime=16861873, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT verdict: Established, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 1/2: src=10.10.0.237:69 dst=192.168.36.12:42671
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT entry found lifetime=16861873, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT verdict: Reply, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: Successfully mapped daddr=192.168.36.12 to identity=6
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 2553 DEBUG: 61 bytes, Delivery to ifindex 0
Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=12:ff:6b:94:bf:0e DstMAC=36:7b:50:98:95:0f EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..27..] Version=4 IHL=5 TOS=0 Length=47 Id=25266 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=59744 SrcIP=10.10.0.237 DstIP=192.168.36.12 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..19..] SrcPort=69(tftp) DstPort=42671 Length=27 Checksum=61399}
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 1/2: src=10.10.0.237:69 dst=192.168.36.12:42671
CPU 01: MARK 0x0 FROM 632 DEBUG: Inheriting identity=0 from stack
CPU 01: MARK 0x0 FROM 632 DEBUG: Attempting local delivery for container id 2553 from seclabel 6
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 1/2: src=192.168.36.12:42671 dst=10.10.0.237:69
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT entry found lifetime=16861873, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT verdict: Established, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT entry found lifetime=16861873, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: CT verdict: Reply, revnat=0
CPU 01: MARK 0x0 FROM 2553 DEBUG: Successfully mapped daddr=192.168.36.12 to identity=6
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 2553 DEBUG: 61 bytes, Delivery to ifindex 0
Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=12:ff:6b:94:bf:0e DstMAC=36:7b:50:98:95:0f EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..27..] Version=4 IHL=5 TOS=0 Length=47 Id=25267 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=59743 SrcIP=10.10.0.237 DstIP=192.168.36.12 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..19..] SrcPort=69(tftp) DstPort=42671 Length=27 Checksum=61399}
```

**Comment by brb** [comment]
Hit the loop again, this time with XDP:
```
/// REQUEST 1


CPU 01: MARK 0x0 FROM 0 DEBUG: 86 bytes, Unknown message type=1 arg1=667
Ethernet	{Contents=[..14..] Payload=[..78..] SrcMAC=08:00:27:3a:28:06 DstMAC=08:00:27:dc:2e:0e EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..52..] Version=4 IHL=5 TOS=0 Length=72 Id=48676 Flags=DF FragOffset=0 TTL=62 Protocol=UDP Checksum=36309 SrcIP=192.168.36.11 DstIP=10.10.1.238 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..44..] SrcPort=46731 DstPort=69(tftp) Length=52 Checksum=20398}
CPU 01: MARK 0x0 FROM 3733 DEBUG: Inheriting identity=2 from stack
CPU 01: MARK 0x0 FROM 3733 DEBUG: Attempting local delivery for container id 855 from seclabel 6
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=192.168.36.11:46731 dst=10.10.1.238:69
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: New, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack create: proxy-port=0 revnat=0 src-identity=6 lb=0.0.0.0
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 855 to-endpoint: 86 bytes (86 captured), state new, interface lxc68960c70d0e2, identity 6->7452, orig-ip 192.168.36.11, to endpoint 855
Ethernet	{Contents=[..14..] Payload=[..78..] SrcMAC=6e:74:58:fd:9c:f8 DstMAC=3a:cb:c3:80:62:38 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..52..] Version=4 IHL=5 TOS=0 Length=72 Id=48676 Flags=DF FragOffset=0 TTL=60 Protocol=UDP Checksum=36821 SrcIP=192.168.36.11 DstIP=10.10.1.238 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..44..] SrcPort=46731 DstPort=69(tftp) Length=52 Checksum=20398}


/// REPLY 1


CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=10.10.1.238:69 dst=192.168.36.11:46731
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x0 FROM 855 DEBUG: CT entry found lifetime=16818314, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: Reply, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Successfully mapped daddr=192.168.36.11 to identity=6
CPU 01: MARK 0x0 FROM 855 DEBUG: Local container found ifindex 668 seclabel 256
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 855 to-stack: 66 bytes (66 captured), state reply, identity 7452->6, orig-ip 0.0.0.0
Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=3a:cb:c3:80:62:38 DstMAC=6e:74:58:fd:9c:f8 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..32..] Version=4 IHL=5 TOS=0 Length=52 Id=34672 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=50077 SrcIP=10.10.1.238 DstIP=192.168.36.11 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..24..] SrcPort=69(tftp) DstPort=46731 Length=32 Checksum=61660}
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 855 DEBUG: 66 bytes, Delivery to ifindex 666
Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=3a:cb:c3:80:62:38 DstMAC=6e:74:58:fd:9c:f8 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..32..] Version=4 IHL=5 TOS=0 Length=52 Id=34672 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=50077 SrcIP=10.10.1.238 DstIP=192.168.36.11 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..24..] SrcPort=69(tftp) DstPort=46731 Length=32 Checksum=61660}


/// WHO IS THIS?


CPU 01: MARK 0x0 FROM 3733 DEBUG: Inheriting identity=0 from stack
CPU 01: MARK 0x0 FROM 3733 DEBUG: Attempting local delivery for container id 855 from seclabel 6
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=192.168.36.11:46731 dst=10.10.1.238:69
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT entry found lifetime=16818314, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: Established, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=10.10.1.238:69 dst=192.168.36.11:46731
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x0 FROM 855 DEBUG: CT entry found lifetime=16818314, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: Reply, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Successfully mapped daddr=192.168.36.11 to identity=6
CPU 01: MARK 0x0 FROM 855 DEBUG: Local container found ifindex 668 seclabel 256
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 855 DEBUG: 140 bytes, Delivery to ifindex 666
Ethernet	{Contents=[..14..] Payload=[..118..] SrcMAC=3a:cb:c3:80:62:38 DstMAC=6e:74:58:fd:9c:f8 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..98..] Version=4 IHL=5 TOS=0 Length=126 Id=34673 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=50002 SrcIP=10.10.1.238 DstIP=192.168.36.11 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..90..] SrcPort=69(tftp) DstPort=46731 Length=106 Checksum=61734}
  Packet has been truncated


CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=10.10.1.238:69 dst=192.168.36.11:46731
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=1
CPU 01: MARK 0x0 FROM 855 DEBUG: CT entry found lifetime=16818314, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: Reply, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Successfully mapped daddr=192.168.36.11 to identity=6
CPU 01: MARK 0x0 FROM 3733 DEBUG: Inheriting identity=0 from stack
CPU 01: MARK 0x0 FROM 3733 DEBUG: Attempting local delivery for container id 855 from seclabel 6
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 1/2: src=192.168.36.11:46731 dst=10.10.1.238:69
CPU 01: MARK 0x0 FROM 855 DEBUG: Conntrack lookup 2/2: nexthdr=17 flags=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT entry found lifetime=16818314, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: CT verdict: Established, revnat=0
CPU 01: MARK 0x0 FROM 855 DEBUG: Local container found ifindex 668 seclabel 256
------------------------------------------------------------------------------
CPU 01: MARK 0x0 FROM 855 DEBUG: 61 bytes, Delivery to ifindex 666
Ethernet	{Contents=[..14..] Payload=[..54..] SrcMAC=3a:cb:c3:80:62:38 DstMAC=6e:74:58:fd:9c:f8 EthernetType=IPv4 Length=0}
IPv4	{Contents=[..20..] Payload=[..27..] Version=4 IHL=5 TOS=0 Length=47 Id=34674 Flags=DF FragOffset=0 TTL=63 Protocol=UDP Checksum=50080 SrcIP=10.10.1.238 DstIP=192.168.36.11 Options=[] Padding=[]}
UDP	{Contents=[..8..] Payload=[..19..] SrcPort=69(tftp) DstPort=46731 Length=27 Checksum=61655}
```

**Comment by brb** [comment]
Hmm, the loop might be explained by https://github.com/cilium/cilium/issues/12241.

**Comment by nebril** [comment]
@brb do you think this was fixed then? Let's monitor.

**Comment by brb** [comment]
Unfortunately, there is one more case - https://github.com/cilium/cilium/issues/8945#issuecomment-635342529.

**Comment by jrajahalme** [comment]
> Unfortunately, there is one more case - [#8945 (comment)](https://github.com/cilium/cilium/issues/8945#issuecomment-635342529).

@brb @nebril  Fixes of two bugs that have caused these errors have been merged since all of these cases:

- #12248 (Fixes #12241) - socket match rule matched all UDP packets destined to the DNS proxy listening port regardless of source or destination IP address (or source port). Replies were lost if the source port number of the original request on one node was the same as the DNS proxy listening port on another node.
- #11899 - NOTRACK rule matched hostport traffic not destined to lxc+ or cilium_host. Replies from the ingress proxy were not reverse NATted and hence were sent to the requesting node with incorrect source IP address and port, causing the original direction source node to respond with an RST.

If the failing test was not using L7 proxy redirection (e.g., no L7 policy, or TFTP with L7 policy (on http)), the problem was the first one. For L7 redirected http traffic the problem was most likely the 2nd one, but less likely also the 1st one.

**Comment by jrajahalme** [comment]
Fixed by #12248 and/or #11899, closing.
